tools/install-powershell.ps1

```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
     }
 }
 
+Function IsPathInRegistryPath([string] $Path) {
+    #Remove ending DirectorySeparatorChar for comparison purposes
+    $Path = [System.Environment]::ExpandEnvironmentVariables([regex]::Replace($Path, '\\;$|\\$', ''));
+    $InstalledPaths = @()
+    #Possible gotcha: don't use Get-ItemPropertyValue (see https://github.com/PowerShell/PowerShell/issues/5906)
+    #Possible gotcha: coerce the (possibly null) property to string before calling ExpandEnvironmentVariables
+    [string] $HKCURegKeyPath = Get-ItemProperty -Path 'HKCU:\Environment' -Name 'PATH' -ErrorAction SilentlyContinue
+    $InstalledPaths += @([System.Environment]::ExpandEnvironmentVariables(($HKCURegKeyPath)).Split(';'))
+    [string] $HKLMRegKeyPath = Get-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\Environment' -Name 'PATH' -ErrorAction SilentlyContinue
+    $InstalledPaths += @([System.Environment]::ExpandEnvironmentVariables(($HKLMRegKeyPath)).Split(';'))
+    #Remove ending DirectorySeparatorChar in all items of array for comparison purposes
+    $InstalledPaths = $InstalledPaths.ForEach( { $_.Replace('\\$', '') })
+    if ($InstalledPaths -icontains $Path) {
+        return $True
+    }
+    return $False
+}
+
+Function SavePathInRegistryKey([Microsoft.Win32.RegistryKey] $Key, [string] $Path) {
+    [string] $CurrentUnexpandedValue = $Key.GetValue('PATH', '', [Microsoft.Win32.RegistryValueOptions]::DoNotExpandEnvironmentNames)
+    #Keep current PathValueKind if possible/appropriate
+    try {
+        [Microsoft.Win32.RegistryValueKind]$PathValueKind = $Key.GetValueKind('PATH')
+    } catch {
+        [Microsoft.Win32.RegistryValueKind]$PathValueKind = [Microsoft.Win32.RegistryValueKind]::ExpandString
+    }
+    #Evaluate new path
+    $NewPathValue = [string]::Concat(
+        [regex]::Replace($CurrentUnexpandedValue, ([System.IO.Path]::DirectorySeparatorChar + '$')),
+        [System.IO.Path]::DirectorySeparatorChar, $Path)
+    #Upgrade PathValueKind to [Microsoft.Win32.RegistryValueKind]::ExpandString if appropriate
+    if ($NewPathValue.Contains('%')) { $PathValueKind = [Microsoft.Win32.RegistryValueKind]::ExpandString }
+    $Key.SetValue("PATH", $NewPathValue, $PathValueKind)
+}
+
+Function AddPathToPathWinEnv([string]$Path) {
+    if (-not (IsPathInRegistryPath($Path))) {
+        #Do not save user paths in machine wide registry path
+        #USERPROFILE is not a subpath of APPDATA or LOCALAPPDATA for roaming profiles
+        #APPDATA and LOCALAPPDATA are relocatable ==> test all 3
+        if ((-not ($Path.StartsWith($ENV:USERPROFILE))) -and
+            (-not ($Path.StartsWith($ENV:APPDATA))) -and
+            (-not ($Path.StartsWith($env:LOCALAPPDATA)))) {
+            try {
+                [string]$KeyName = 'SYSTEM\CurrentControlSet\Control\Session Manager\Environment\'
+                SavePathInRegistryKey([Microsoft.Win32.Registry]::LocalMachine.OpenSubKey($KeyName), $Path)
```

Comment:
Please use Environment APIs
```suggestion
[System.Environment]::SetEnvironmentVariable('PATH',$Path,'Machine')
```

Reply:
but.... this would reintroduce the problems this PR is trying to avoid!  may as well close the pr then.

Reply:
How so, these two do the same thing but with the later it is more clear what your are doing?

Reply:
I review.  You should really file an issue on DotNet as well.  You shouldn't have to resort to Registry calls to alter the environment.

Reply:
### TLDR: nobody wants to touch this with a 10 foot clown pole.

> You should really file an issue on DotNet as well

I'll write a solid draft of everything I know about this, links and all, if you are willing to send a proposal.

### How the windows api team sees this (as imagined by me)

I like the way the windows api sees things myself.  Their view on the matter (this is all interpretation on my part from what the api does) is that there is no such thing as an _environment variable_ outside of a process.  [The docs clearly state that a call to SetEnvironmentVariable changes the process environment block (PEB) only](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-setenvironmentvariable).  They offer completely different methods for saving values in the registry.

### Design issues with SetEnvironmentVariable

The dotnet team tried to _fuzz_ the distinction between a _variable-living-inside-a-process_ and a _setting-whose-value-is-set-in-the-registry_.  The ["target" parameter](https://referencesource.microsoft.com/mscorlib/a.html#821defd26e0b02d7) (which does not exist in the windows api embodies a perfect example of a single responsibility principle violation.

It's possible values are process, user and machine.  The first one implies changes in memory while the other two are clearly about changing values in the registry.   

### Implementation details in SetEnvironmentVariable
For the 2 targets that end up targeting the registry, i.e. USER and MACHINE,
[SetEnvironmentVariable relies on Microsoft.Win32.Registry](https://referencesource.microsoft.com/mscorlib/a.html#33fd9bc66f5473b7) class to implement registry persistence.  

Because SetEnvironmentVariable doesn't provide an overload with a RegistryValueKind tyoe parameter however, it defaults to saving RegistryValueKind.REG_SZ, disabling expansion on any %VARS% in the path.  Cutting out the middle man (SetEnvironmentVariable) and and relying on Microsoft.Win32.Registry is the only option available.

### What would need to happen for a fix in Environment
- All things considered, the easiest path seems to be to add an overload to SetEnvironmentVariable  that would accept a RegistryValueKind as a parameter.
- Adding a RegistryValueKind parameter will surely bring requests to support REG_BINARY, REG_MULTI_SZ and others as well, meaning more overloads (one with a value parameter that is of type Byte[], that will persist as a REG_BINARY, one with a value parameter of type Int that will persist as a REG_DWORD and implement auto-detection of REG_MULTILINE_SZ).

### Why I think any proposal coming from outside Microsoft won't be fixed?
The payoff is small (set the value of two miserable registry values). and the other reason isn't very compelling = we're doing this so users don't have to think about the difference between a registry value and variables living in a PEB.

### Why I think you might have a chance
In theory the golden rule that asks everyone to **always assume competence** from the person that wrote the PR or the bug report or the code review is correct.  A few years of badly written PRs, angry and/or  disrespectful bug reports can't help but to quickly chip away at this golden rule idea however.  A PR coming from a colleague has a chance to be taken seriously.

Reply:
(no changes to code)